### DISCUSSION TITLE
My 2 LT / Modbus / Cascade installation

### DISCUSSION THREAD

**Discussion by gvzdus** [Egyras/HeishaMon#565] [Show and tell]
As geduxas asked, I think it is not pretentious to present my installation.

**Why that complex, expensive shit?!**
Preference was more on "getting a new toy" (that's, why I chose Panasonic because of Heishamon) as well as staying married, rather than economy. Because of 5% more german subsidy, I wanted an R290 solution, and as I need 18 kW heating peak power, I chose a dual heatpump system.

**Now, what in detail do you have?**
2 x Aquarea LT Generation L WH-WDG09LE5 + WH-SDC0509L3E5 + CZ-TAW1B (Wifi Modul)
2 x A2W CZ-NS5P (extension board)
2 x PAW-AW-MBS-H (modbus interface)
2 x A2W CZ-NE4P (heating)
1 x PAW-A2W-CMH (cascade manager)
The system feeds a buffer tank ("Capito"), about 550l for DHW and 200l buffer for heating (configured as parallel buffer)
Sensors include outdoor temp, DHW buffer temp and heating buffer, but no indoor sensors.

**... and no Heishamon?**
Of course!! 2 large boards + Ethernet, delivered superfast by TheHogNL.
I decided for now, that I keep the CZ-TAW1B off (pretty limited use, unhappy to install additional USB power supply), and keep the "sniffer" Modbus Raspi in the system. The Heishamons work great and feed their data to a mosquitto broker on my OpenWB wallbox, where I process the data in my smarthome system (FHEM). Currently, I use the Modbus-Raspi also via FHEM, to emulate 3 Siemens energy meters: The cascade manager has a "PV optimization", that requests 3 Siemens Modbus energy meters to deliver the power of (1: PV, 2: home consumption, 3: heatpump consumption). The manager simply asks for one modbus register, and I simulate this from my smarthome system.

**Lessons learned so far?**
The system is running since 2024-09-02, so not that long. The cascade manager is stupid pain:
- Worst thing: No error messages. Just blinking red, if unhappy with a heatpump. No "Modbus comm failure", no "your setting in the heatpump is wrong" message. My installer forgot to configure "optional pcb installed" on heatpump 2, and he configured a heat curve instead of fixed temperature on hp 2, which is not right for the cascade manager (but the installation manual does not tell this). While the cascade manager polls tons of documented and undocumented modbus registers all few seconds, it does not issue warnings about these misconfigurations. I have debugged with a Raspi & a snooping Modbus interface (Modbus RTP, not TCP).
- No "Update" option
- No "System log": "I have started HP 2 because of ..."
- Pretty limited Modbus/TCP registers of the cascade manager - Heishamon tells a lot more!
- Undocumented behaviour, few parameters for e.g. PV optimization.

By having a parallel buffer, I do loose about 3K in the buffer. E.g.: Heating system is working at 27° in/30° out, heatpump is operating with 30° in & 33° out. That might be about 10% efficiency loss.

**How can you help & contribute?**
I am not the hardware guy, more the Java & Linux ops guy. I sweat when I am asked to flash a firmware on an ESP32. But I would be willing to "do guided steps on hardware". Or send debug logs if told how.
Email preferred (gvz is the user part, garnix o de my domain) to limit public visibility.

**And no pictures??!**
Here is one. Modbus interface on top, in the bottom Heishamon board & Raspi.
<img width="597" alt="image" src="https://github.com/user-attachments/assets/78de6c5d-c5c9-4cf9-a602-a85314cecf8d">

**Comment by gvzdus** [comment]
Just as a suggestion: If you want to introduce your solution, start a new thread!
Let's keep the comments for Q&A or some kind of "personal message".

**Comment by geduxas** [comment]
Nice setup! I am also interested more in Modbus interface analysis, maybe there we could find something more about protocol.. 
As for modbus adapter you could easily emulate protocol of the pump. 

Also about firmware upgrade, don't be so scered, it's easy task.. it's almost same as reinstalling OS  :D

**Reply by gvzdus** [reply]
Really willing to help, but I'd prefer email, as I e.g. do not want to upload documents that might infringe copyright issues.

**Comment by gvzdus** [comment]
What I can state and publish, is what I see:
The cascade manager takes HP1 as the leading HP, where parameters (e.g. sensor values, setting from the cloud) are taken from.
But how does it prevent HP1 from starting and enforces HP2 instead? (It seems to try to start the HP that has less op. hours or op. starts).
It writes to HP1 via Modbus ("Stay off"):
#20:25:34.135946 U01 WriteReg(152) = 1
#20:25:34.177634 U01 WriteReg(154) = 0
#20:25:34.218431 U01 WriteReg(156) = 100
#20:25:34.258566 U01 WriteReg(157) = 100

The active HP gets via Modbus:
21:53:25.921471 U02 WriteReg(0) = 1
21:53:25.974457 U02 WriteReg(4) = 1
21:53:26.011171 U02 WriteReg(12) = 36
21:53:26.053170 U02 WriteReg(33) = 40

Documented registers are:
0 = System On / Off
4 = Op mode (1 = heat)
12 = Zone Z1 setpoint temperature 
33 = Tank water setpoint temperature
So, all registers to the active heat pump are documented, all registers to "suppress" the leading heatpump from starting are undocumented

**Reply by geduxas** [reply]
How often controller sending write commands to pump? Do you have such statistics?

**Comment by gvzdus** [comment]
Damn often.
The full 135 registers are read all 3-4 seconds for each unit.
The Writes are repeated all 20-30 seconds.

Btw: Just took the cascade offline again: It kept on telling to set the target temp to +5K, although PV was at 0, and although I had disabled PV optimization as first step...

**Reply by geduxas** [reply]
I am asking about writes, because there is some speculation about EPROM lifetime.. it would be great to know how much writes doing official Panasonic tool's. 

Aa for cascade, i think it's logic is not a logical in some cases.. we see some sort not logical works in pumps also..  also it's priority i think is heating, or you have some setup/configuration error's

**Comment by gvzdus** [comment]
It *might* be that the PAW-AW-MBS-H "caches" the Writes (better: only passes them to the internal bus if it knows that they are changes). But I don't think so...

**Comment by gvzdus** [comment]
Just to complete the description: Here's the diagram:
<img width="1111" alt="image" src="https://github.com/user-attachments/assets/1d30c3ff-4dd0-466c-a8b1-0e50e9e7560e" />
